federation/pkg/federation-controller/service/dns.go

```diff
@@ -16,25 +16,244 @@ limitations under the License.
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		for _, ingress := range lbStatus.Ingress {
+			var address string
+			// We should get either an IP address or a hostname - use whichever one we get
+			if ingress.IP != "" {
+				address = ingress.IP
+			} else if ingress.Hostname != "" {
+				address = ingress.Hostname
+			}
+			if len(address) <= 0 {
+				return nil, nil, nil, fmt.Errorf(
+					"Service %s/%s in cluster %s has neither LoadBalancerStatus.ingress.ip nor LoadBalancerStatus.ingress.hostname."+
+						"Cannot use it as endpoint for federated service.",
+					cachedService.lastState.Name, cachedService.lastState.Namespace, clusterName)
+			}
+			for _, lbZoneName := range lbZoneNames {
+				for _, zoneName := range zoneNames {
+					if lbZoneName == zoneName {
+						zoneEndpoints = append(zoneEndpoints, address)
+					}
+				}
+			}
+			if lbRegionName == regionName {
+				regionEndpoints = append(regionEndpoints, address)
+			}
+			globalEndpoints = append(globalEndpoints, address)
+		}
+	}
+	return zoneEndpoints, regionEndpoints, globalEndpoints, nil
 }
 
-// getClusterRegionName returns the name of the region where the specified cluster exists (e.g. us-east1 on GCE, or "us-east-1" on AWS)
-func getClusterRegionName(clusterName string) string {
-	// TODO: quinton: Get this from the cluster API object - from the annotation on a node in the cluster - it doesn't contain this yet.
-	return "region-of-cluster-" + clusterName
+// getClusterZoneNames returns the name of the zones (and the region) where the specified cluster exists (e.g. zones "us-east1-c" on GCE, or "us-east-1b" on AWS)
+func (s *ServiceController) getClusterZoneNames(clusterName string) (zones []string, region string, err error) {
+	if client, ok := s.clusterCache.clientMap[clusterName]; !ok {
+		return nil, "", fmt.Errorf("Cluster cache does not contain entry for cluster %s", clusterName)
+	} else {
+		if client.cluster == nil {
+			return nil, "", fmt.Errorf("Cluster cache entry for cluster %s is nil", clusterName)
+		} else {
+			return client.cluster.Status.Zones, client.cluster.Status.Region, nil
+		}
+	}
 }
 
 // getFederationDNSZoneName returns the name of the managed DNS Zone configured for this federation
-func getFederationDNSZoneName() string {
-	return "mydomain.com" // TODO: quinton: Get this from the federation configuration.
+func (s *ServiceController) getFederationDNSZoneName() (string, error) {
+	return "example.com", nil // TODO: quinton: Get this from the federation configuration.
+	// Note: For unit testing this must match the domain populated in the test/stub dnsprovider.
 }
 
-func ensureDNSRecords(clusterName string, cachedService *cachedService) error {
-	// Quinton: Pseudocode....
+// getZone is a hack around the fact that dnsprovider does not yet support a Get() method, only a List() method.  TODO:  Fix that.
+func getZone(dnsZoneName string, zonesInterface dnsprovider.Zones) (dnsprovider.Zone, error) {
```

Comment:
I did that because it's explicitly wrapping an interface method Zone().  I didn't think it made sense to call the wrapper function getDnsZone().  And as you say, it's in a file called dns.go, so it shouldn't be too confusing.  I don't fell very strongly about it, and can change it in a followup PR if you like.